Jenkinsfile

```diff
@@ -96,36 +97,15 @@ def job = {
         },
         downstreamBuildsStepName: {
             echo "Building cp-downstream-builds"
-            if (config.isPrJob) {
-                try {
-                    def muckrakeBranch = kafkaMuckrakeVersionMap[env.CHANGE_TARGET]
-                    def forkRepo = "${env.CHANGE_FORK ?: "confluentinc"}/kafka.git"
-                    def forkBranch = env.CHANGE_BRANCH
-                    echo "Schedule test-cp-downstream-builds with :"
-                    echo "Muckrake branch : ${muckrakeBranch}"
-                    echo "PR fork repo : ${forkRepo}"
-                    echo "PR fork branch : ${forkBranch}"
-                    buildResult = build job: 'test-cp-downstream-builds', parameters: [
-                            [$class: 'StringParameterValue', name: 'BRANCH', value: muckrakeBranch],
-                            [$class: 'StringParameterValue', name: 'TEST_PATH', value: "muckrake/tests/dummy_test.py"],
-                            [$class: 'StringParameterValue', name: 'KAFKA_REPO', value: forkRepo],
-                            [$class: 'StringParameterValue', name: 'KAFKA_BRANCH', value: forkBranch]],
-                            propagate: true, wait: true
-                    downstreamBuildFailureOutput = "cp-downstream-builds result: " + buildResult.getResult();
-                    return downstreamBuildFailureOutput
-                } catch (Exception e) {
-                    currentBuild.result = 'UNSTABLE'
-                    downstreamBuildFailureOutput = "cp-downstream-builds result: " + e.getMessage()
-                    writeFile file: "downstream/cp-downstream-build-failure.txt", text: downstreamBuildFailureOutput
-                    archiveArtifacts artifacts: 'downstream/*.txt'
-
-                    return downstreamBuildFailureOutput
+            stage('Downstream validation') {
+                if (config.isPrJob && config.downStreamValidate) {
+                    downStreamValidation(true)
```

Comment:
When the downstream validation runs and fails downstream how will the build fail? Will it be UNSTABLE or will it be FAILURE?

Reply:
It will try to compile all downstream repos, as long as one fail(might be multiple failures), the results will be FAILURE.

Reply:
I would like to avoid reintroducing this behavior, as we previously changed it to mark downstream build failures as UNSTABLE due to PRs being merged that break master.

What ends up happening is something like the following:
1. Downstream breaks.
2. PRs which weren't the cause of the failure end up being marked as FAILED, and so people have to check whether it was due to an existing downstream failure or their PR.
2. Eventually a FAILED build comes in which was due to changes in the PR (say it failed to compile and thus there were also no unit test failures either) while downstream is actually OK. Thinking the FAILED build was just due to master, they go ahead and merge their PR, breaking master. There are other issues like gradle workers exiting with non-zero exit codes that also lead to a red build, and we end up losing these strong signals in the noise of builds being marked as FAILED as a result of the downstream failures.

Reply:
Thanks for the response, this is a draft PR now, I am still testing various stuff to make sure it does not break kafka's current workflow. For your comments, I have few questions and clarifications:

> Downstream breaks.

https://confluentinc.atlassian.net/wiki/spaces/TOOLS/pages/1179166502/Dependency-mgmt+FAQ, you can disable the check with replay once you confirm the downstream build is broken.(might be little trouble since it require VPN to access it)

> PRs which weren't the cause of the failure end up being marked as FAILED, and so people have to check whether it was due to an existing downstream failure or their PR.

I may not familiar with the kafka workflow, either the build is marked as FAILED or UNSTABLE I think people should check which part become FAILED or UNSTBALE anyway?

> while downstream is actually OK. Thinking the FAILED build was just due to master, they go ahead and merge their PR, breaking master.

I assume the situation here is we know the PR will break downstreams but we have to do it, in such case, we should also disable the downstream checks.

I understand that go into Jenkins, use `replay` to disable the check might be trouble in workflow, though I think downstream failures should be treated little bit different than flaky. But I will follow the kafka team's idea.

Reply:
@xli1996 sounds good, re: testing.

The unfortunate reality is that the downstream builds end up failing pretty often and then all of our PRs become red (failed) even if the PR in question is not to blame. Then things get merged which break master because it's difficult to pick out the cause of the failures in the noise. I agree that we should be checking the cause of the failure, but the reality of it is that it's very tricky to do so and marking the whole build as failed is pretty extreme given that the downstream build is broken and it is not even due to the PR in question. Anyway, just wanted to surface that it's a concern for me, as this behavior being reintroduced in this PR lead to a lot of breakage.

Reply:
I understand your concern, and I try to make downstream failure very easy to detect, you don't need to go to the job logs to search what is failed. (one failure example here https://jenkins.confluent.io/job/confluentinc-pr/job/common/job/PR-255/)

I think I will temporary keep it as FAILURE first, but if you feel it troubles at anytime, we can change it immediately.